### ISSUE TITLE
Consider an alternative disk representation for datom flags

### ISSUE THREAD

**Issue by rnewman** [mozilla/datomish#29]
We have four fields in `datoms`: `index_vaet`, `index_avet`, `index_fulltext` and `unique_value`.

These fields appear in `datoms` and in the indices `idx_datoms_eavt` and `idx_datoms_aevt`.

Some rows also appear in `idx_datoms_avet`, `idx_datoms_vaet`,  .

These fields

**Comment by rnewman** [comment]
Note that each of these columns also appears transiently when inserting datoms into `tx_lookup`, so this has a bearing on #28, too.

**Comment by rnewman** [comment]
SQLite 3.15 allows the use of deterministic functions in partial index expressions, so we can now go whole-hog on bitfields if we choose.

https://github.com/mapbox/node-sqlite3/pull/724 adds this for `node-sqlite3`.

**Comment by rnewman** [comment]
Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1312455.

**Comment by rnewman** [comment]
Hooray: https://bugzilla.mozilla.org/show_bug.cgi?id=1310361 already landed.

**Comment by rnewman** [comment]
`0.2.0-SNAPSHOT` now includes SQLite 3.15.

**Comment by thomcc** [comment]
I did this in https://github.com/thomcc/mentat/commit/ba29de5998d8a905e024339e9e5167ee56f2551c (note that not all tests pass, but it was just a test to see how beneficial it was). Sadly, it does not appear to make a meaningful difference in the on-disc database size :(

**Comment by rnewman** [comment]
You need to make `unique` the least significant bit. It’s common, and here it guarantees you a byte, so versus separate-zeroes it’ll be of little benefit.

**Comment by thomcc** [comment]
Yeah, it's not nothing (average of 2-3 bytes per datom), but not really clear that it's worth the version breakage or additional complexity. It probably saves more for things using `:db/index true` though!

```
$ ls -al
<snip>
-rw-r--r--   1 thom  staff  194564096 Aug 20 20:50 mentat_places_with_flag_opt.db
-rw-r--r--   1 thom  staff  197406720 Aug 20 20:50 mentat_places.db
<snip>
$ sqlite3 mentat_places.db
SQLite version 3.19.3 2017-06-27 16:48:08
Enter ".help" for usage hints.
sqlite> select count(*) from datoms;
1119529
$ node
> 197406720 - 194564096
2842624
> (197406720 - 194564096) / 1119529
2.5391249355755856
```
